turbinia/jobs/artifact_extraction.py

```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""job to conditionally extract artifacts from a filesystem image"""
+
+from __future__ import unicode_literals
+
+from turbinia.evidence import Directory
+from turbinia.evidence import GoogleCloudDisk
+from turbinia.evidence import GoogleCloudDiskRawEmbedded
+from turbinia.evidence import ExportedFileArtifact
+from turbinia.evidence import RawDisk
+from turbinia.evidence import ReportText
+from turbinia.jobs import interface
+from turbinia.jobs import manager
+from turbinia.workers.artifact import FileArtifactExtractionTask
+
+
+class ArtifactExtractionJob(interface.TurbiniaJob):
+  """ Extract artifacts specificed in the evidence recipe """
+  evidence_input = [
+      Directory, RawDisk, GoogleCloudDisk, GoogleCloudDiskRawEmbedded
+  ]
+
+  evidence_output = [ExportedFileArtifact]
+  NAME = "ArtifactExtractionJob"
+
+  def create_tasks(self, evidence):
+    """Create task.
+
+    Args:
+      evidence: List of evidence objects to process
+
+    Returns:
+        A list of tasks to schedule.
+    """
+    #Retrieve appropriabe base config
+    task_recipe = self.evidence.config.get('FileArtifactExtractionTask', None)
```

Comment:
I think that if possible we shouldn't implement applying the recipe/variant in the Job, but rather in the Task.  I'm hoping that from the Job/task manager standpoint, that applying the recipe/variant is transparent so that we don't have to change all of the Task creation in every Job to account for this.

Right now the recipe data is put into the request by the client and is then plumbed through to the Task and is available as `evidence.config` from the `run()` method.  I think we might want to change the `BaseTaskManager.add_task()` method to just set the `task.recipe` to whatever is in `evidence.config` before it passes it to `enqueue_task()`, and that way we'll always have available at the Task.

Once the recipe data is available in the Task, then in `TaskFoo.run()` can just do things like:
```
artifacts = self.recipe.get('artifacts', [FooArtifact])
```

Side note: I hadn't actually thought much about reading/applying recipes at the Job layer, but I could imagine that there are parameters that could affect `create_tasks()` in some circumstances, but I think that should be the exception when needed.

Reply:
I am inclined to think that delegating the task generation to each instance of create_tasks is a more 'transparent' way of handling the creating of specific tasks, with add_task simply doing the job of adding a task and not caring about the logic behind creating one (or not). Incorporating recipes will bring about reviewing the code for each worker too, which is a much more involved change. For virtually all jobs, abstracting away some logic such as the following in the TurbiniaJob class, would fit the bill.

```python
   task_recipe = self.evidence.config.get(task_name, None)
    if task_recipe:
      tasks = []
      for variant in task_recipe['variant']:
        for _ in evidence:
          new_task = TaskClass(task_variant=variant)
          if self.validate_task_conf(new_task.task_conf, task_recipe[variant]):
            tasks.append(new_task)
    else:
      tasks = [TaskClass() for _ in evidence]
    return tasks
```

if it is decided to put the logic in add_tasks instead. Should we add a 'NAME' attribute to TurbiniaTask to serve as an anchor point for the function to get to the correct key in the recipe dictionary?

Reply:
Come to think about it. Assuming the job serves no purpose other than executing the stock task creation routine that could be place in the base Job class. Could we consider having tasks be created without matching jobs and have code that creates, initialises and registers jobs for these tasks instead? If this were possible, creating a task that Turbinia could leverage would be a bit simpler too. thoughts?

Reply:
Couple thoughts here:

* Regarding Jobs and whether `create_tasks()` is even needed:  I would still like to have the flexibility of having a Job that spawns multiple configurable Tasks that are calculated by the Job (e.g. breaking up a disk image to block ranges and then having separate Tasks configured to process each range), but you're right that for 90% of the Jobs those methods are all the same, so I think your suggest of having a generic `create_tasks()` as part of the base module would work, but I think we should do that in a separate PR.

* Regarding variants: I guess I was originally thinking about recipes with variants a little differently as more of a easy way to select from a different set of configuration options for a given Task and not necessarily as a way to create a series of differently configured Tasks.  I think this could be considered Job configuration vs. Task configuration.  Maybe we should have a quick chat about how variants can be used, and what the use-cases are?  I'm wondering if many of the use cases could be covered by just rearranging parts of the recipe instead. 

* If we restrict the recipe to running one variant at a time, then we don't have to expose the fact that there are variants at all to the underlying Jobs/Tasks, and then they can just read the Recipe data as general configuration to simplify things.  Doing this looks like it's a little different than you were thinking though, so it's another thing we could talk about.